### PR TITLE
Fixes to allow Configuration Caching in Gradle 6.6+

### PR DESCRIPTION
**Pull request by timyates**
We (myself and @britter) believe this PR allows the Shadow Plugin to be used
in a Gradle 6.6+ build with --configuration-cache enabled.

We unfortunately did not get round to fixing the case where `minimize()` is
used, so it does not work if that is called in the build.

We're basically looking for suggestions on what we can do to improve it,
or whether it is good enough to go in as-is

Cheers

Tim (and Benedikt)

### CODE REVIEW

**Comment by timyates** [comment]
(upcoming configuration cache feature is described here https://docs.gradle.org/release-candidate/userguide/configuration_cache.html)

**Comment by eskatos** [comment]
@timyates, @britter, any chance you could add a failing test for `minimize()`? I think this could help moving this forward

**Comment by britter** [comment]
@eskatos I think you only need to call `minimize()` in the `ShadowJar` configuration. Haven't checked it, but this test case should fail:

```groovy
    def "configuration caching supports minimize"() {
        given:
        buildFile << """
            shadowJar {
               minimize()
            }
        """.stripIndent()

        when:
        runner.withArguments('--configuration-cache', 'shadowJar').build()
        output.delete()
        def result = runner.withArguments('--configuration-cache', 'shadowJar').build()

        then:
        contains(output, ['a.properties', 'b.properties'])

        and:
        doesNotContain(output, ['a2.properties'])
        result.output.contains("Reusing configuration cache.")
    }
```

**Comment by eskatos** [comment]
With the latest changes above, the `minimize {}` use case now works with the configuration cache enabled.

@johnrengelman this should be ready for review, could you please take a look?

BTW it looks like the circleci build is broken
https://app.circleci.com/pipelines/github/johnrengelman/shadow/42/workflows/1b019a05-d7a1-47e5-aa42-3a45781b228e/jobs/497

File: src/main/groovy/com/github/jengelman/gradle/plugins/shadow/ShadowApplicationPlugin.groovy [hunk 1]
```diff
@@ -54,12 +54,10 @@ class ShadowApplicationPlugin implements Plugin<Project> {
...
     protected void configureJarMainClass(Project project) {
-        ApplicationPluginConvention pluginConvention = (
-                ApplicationPluginConvention) project.convention.plugins.application
-
-        jar.inputs.property('mainClassName', { pluginConvention.mainClassName })
+        def classNameProvider = project.provider { project.convention.plugins.application.mainClassName }
+        jar.inputs.property('mainClassName', classNameProvider)
         jar.doFirst {
-            manifest.attributes 'Main-Class': pluginConvention.mainClassName
+            manifest.attributes 'Main-Class': classNameProvider.get()
```
Comment (vlsi) [inline src/main/groovy/com/github/jengelman/gradle/plugins/shadow/ShadowApplicationPlugin.groovy:60]: manifest.attributes support providers: https://twitter.com/VladimirSitnikv/status/1260952932981313544

**Pull request merged**